File: Tests/KeyScheduleTests.py
# Hack sys.path to get code from the parent folder
import os,sys,inspect
currentdir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
parentdir = os.path.dirname(currentdir)
sys.path.insert(0,parentdir) 
# </hack>

from KeySchedule import KeySchedule
from KeySchedule import InverseKeySchedule
import Constants
import unittest

class TestKeySchedule(unittest.TestCase):
    def test_schedule_init(self):
        # Appendex A.1
        key = [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c]
        schedule = KeySchedule(key)
        
        w0 = 0x2b7e1516 
        k0 = int(schedule.roundKeys[0])
        self.assertEqual(w0, k0, "KeySchedule initialization failed")

        w1 = 0x28aed2a6
        k1 = int(schedule.roundKeys[1])
        self.assertEqual(w1, k1, "KeySchedule initialization failed")
        
        w2 = 0xabf71588 
        k2 = int(schedule.roundKeys[2])
        self.assertEqual(w2, k2, "KeySchedule initialization failed")
        
        w3 = 0x09cf4f3c 
        k3 = int(schedule.roundKeys[3])
        self.assertEqual(w3, k3, "KeySchedule initialization failed")
        
    def test_inverse_schedule(self):
        key = range(16)
        sch = KeySchedule(key)
        inv = InverseKeySchedule(key)

        for i in range(Constants.Nr()):
            j = (Constants.Nr() * Constants.Nb()) - (i*4)
            nx = inv.next()
            self.assertEqual([int(k) for k in sch.roundKeys[j:j+4]], [int(k) for k in nx], "Inverse keys schedule was wrong")

if __name__ == '__main__':
    try:
        unittest.main() 
    except SystemExit:
        # unittest & my IDE don't play nice
        # It raises a SystemExit when it finishes
        pass

File: C++ Implementation/Tests/KeyScheduleTests.cpp

#include "pch.h"
#include "CppUnitTest.h"
#include <vector>

#include "KeySchedule.h"
#include "InverseKeySchedule.h"


using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace Tests
{
	TEST_CLASS(KeyScheduleTests)
	{
	public:

		TEST_METHOD(TestKeyScheduleAt)
		{
			std::vector<uint8_t> key = { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
			KeySchedule sched(key);

			Assert::AreEqual((uint32_t)0x2b7e1516, sched.at(0));
			Assert::AreEqual((uint32_t)0xdb0bad00, sched.at(19));
			Assert::AreEqual((uint32_t)0xc9ee2589, sched.at(41));
		}

		TEST_METHOD(TestKeyScheduleNext)
		{
			std::vector<uint8_t> key = { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
			KeySchedule sched(key);

			std::vector<uint32_t> expected{ 0x2b7e1516, 0x28aed2a6, 0xabf71588, 0x09cf4f3c, 0xa0fafe17, 0x88542cb1 };
			for (int i = 0; i < expected.size(); ++i) {
				Assert::AreEqual(expected.at(i), sched.next());
			}
		}


		TEST_METHOD(TestInverseKeyScheduleAt)
		{
			std::vector<uint8_t> key = { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
			KeySchedule sched = InverseKeySchedule(std::vector<uint8_t>(key));

			Assert::AreEqual((uint32_t)0x2b7e1516, sched.at(0));
			Assert::AreEqual((uint32_t)0xdb0bad00, sched.at(19));
			Assert::AreEqual((uint32_t)0xc9ee2589, sched.at(41));
		}


		TEST_METHOD(TestInverseKeyScheduleNext)
		{
			std::vector<uint8_t> key = { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
			KeySchedule sched = InverseKeySchedule(key);

			std::vector<uint32_t> expected{ 0xd014f9a8,  0xc9ee2589, 0xe13f0cc8, 0xb6630ca6};
			for (int i = 0; i < expected.size(); ++i) {
				Assert::AreEqual(expected.at(i), sched.next());
			}
		}
	};
}
